### PR TITLE
Enabling custom URL for APIM Publisher

### PR DESCRIPTION
**Pull request by wasuradananjith**
## Purpose
Fixes: https://github.com/wso2/product-apim/issues/8677

## Goals
Provide custom URL support for API publisher.

## Approach
Frontporting https://github.com/wso2-support/carbon-apimgt/pull/2147 and https://github.com/wso2-support/carbon-apimgt/pull/2187 by cherry-pick

### CODE REVIEW

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/services/login/idp.jag [hunk 1]
```diff
@@ -45,8 +45,16 @@
...
-    var logoutCallbackUrl = serverUrl + app.context + LOGOUT_CALLBACK_URL_SUFFIX;
+    var loginCallbackUrl = appUtils.getTenantBasedLoginCallBack();
+    if (loginCallbackUrl == null) {
+	    loginCallbackUrl = serverUrl + app.context + LOGIN_CALLBACK_URL_SUFFIX;
+    }
+
+    var logoutCallbackUrl = appUtils.getTenantBasedLogoutCallBack();
+    if (logoutCallbackUrl == null) {
+         logoutCallbackUrl = serverUrl + app.context + LOGOUT_CALLBACK_URL_SUFFIX;
+     }
```
Comment (ruks) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/services/login/idp.jag:56]: slight formatting issue.

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/services/utils.js [hunk 2]
```diff
@@ -31,10 +32,72 @@ var getLoopbackOrigin = function() {
...
+};
+
+var getTenantBasedLogoutCallBack = function() {
+    var tenantDomain = getTenantDomain();
+    var publisherDomainMapping = utils.getTenantBasedPublisherDomainMapping(tenantDomain);
+    if (publisherDomainMapping != null) {
+        if (publisherDomainMapping.get('logout') != null) {
+            return publisherDomainMapping.get('logout');
+        }
+        return "https://"+publisherDomainMapping.get('customUrl') + LOGOUT_CALLBACK_URL_SUFFIX;
```
Comment (ruks) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/services/utils.js:73]: better to add spaces

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java [hunk 3]
```diff
@@ -10557,6 +10557,69 @@ public static Map getTenantBasedStoreDomainMapping(String tenantDomain) throws A
...
+                JSONParser parser = new JSONParser();
+                JSONObject mappings = (JSONObject) parser.parse(content);
+                if (mappings.containsKey(APIConstants.API_DOMAIN_MAPPINGS_PUBLISHER)) {
+                    return (Map) mappings.get(APIConstants.API_DOMAIN_MAPPINGS_PUBLISHER);
+                }
+            }
+        } catch (RegistryException e) {
+            String msg = "Error while retrieving publisher domain mappings from registry";
+            throw new APIManagementException(msg, e);
+        } catch (ClassCastException e) {
```
Comment (ruks) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java:10578]: Better not to catch ClassCastException. Ideally, you should check the instance type and continue.

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/public/pages/index.jag [hunk 4]
```diff
@@ -25,6 +25,10 @@
     var swaggerKey = 'swaggerWorkerInit.js';
     var indexBundle = manifest[indexKey].split("dist/")[1];
     var swaggerBundle = manifest[swaggerKey].split("dist/")[1];
+    var appUtils = require("/services/utils.js");
+    app.context = appUtils.getTenantBasePublisherContext();
+    var appUtils = require("/services/utils.js");
```
Comment (Arshardh) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/public/pages/index.jag:30]: Shall we remove the duplicate entry?

Comment (wasuradananjith) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/site/public/pages/index.jag]: Sure

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/services/utils.js [hunk 5]
```diff
@@ -38,3 +39,65 @@ function getIDPOrigin() {
...
+
+var getTenantBasedLoginCallBack = function() {
+    var tenantDomain = getTenantDomain();
+    var publisherDomainMapping = utils.getTenantBasedPublisherDomainMapping(tenantDomain);
+    if (publisherDomainMapping != null) {
+        if (publisherDomainMapping.get('login') != null) {
+            return publisherDomainMapping.get('login');
+        }
+        return "https://"+publisherDomainMapping.get('customUrl') + LOGIN_CALLBACK_URL_SUFFIX;
+    }else{
```
Comment (praminda) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/services/utils.js:61]: fix formatting

Comment (wasuradananjith) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/services/utils.js:61]: Fixed via https://github.com/wso2/carbon-apimgt/pull/9604/commits/bcfd5cf98bcbb80353a0c437c56ebcd13e36f5b1

**Pull request merged**